### ISSUE TITLE
[RuntimeAsync] Double check OSR in async2 struct instance methods

### ISSUE THREAD

**Issue by jakobbotsch** [dotnet/runtimelab#2814]
OSR might not be saving the struct instance copy made implicitly by the JIT. Double check and fix this if necessary.

**Comment by jakobbotsch** [comment]
These sort of complications make me think that perhaps we should still leave it up to the code generators to create these copies themselves. I think we are going to run into similar problems around EnC and also diagnostics with the implicitly created locals that are hidden from the user.

Leaving it up to IL creators to create these copies would probably also simplify the spec; it is already saying that byrefs are not captured across suspension points, so the "this" argument in struct methods already fall under this rule, hence IL code generators are forced to create the copy if they need access to the instance after suspension points.

Thoughts @agocke @davidwrighton @VSadov?

**Comment by VSadov** [comment]
We would still need some policy from the JIT on what happens when refs are alive across suspension points. 

Ideally the behavior should not end in behavior that is dependent on whether suspension happened or not or on whether JIT is optimizing.

**Comment by VSadov** [comment]
Rewriting `this` accesses to use a local copy captured at first entry into the method is a fairly straightforward rewrite that Roslyn can do. 
We already decided that Roslyn will spill byrefs into locals for cases like `arr[i] += await foo()`. That would be more complex.

If there are no bugs, there should not be IL byrefs that try to be alive across awaits. I think we can guarantee this on Roslyn side.

I just wonder what will happen if there is a bug, in Roslyn, or in some IL rewriting/instrumentation tool. Will JIT just throw `InvalidProgramException` or leave the behavior dependent on whether suspension happens.  
Like some code may never take suspension path and one day it takes that path and then crashes with NRE while accessing nulled out byref.

I think I might prefer not accepting IL that carries byrefs across awaits. It feels a bit more deterministic, even if it may depend on whether the code was actually JIT-ed and perhaps on quality of flow analysis (and thus on JIT version or optimization mode).

**Comment by jakobbotsch** [comment]
> We would still need some policy from the JIT on what happens when refs are alive across suspension points.

I agree, but I think this is orthogonal to this particular question -- we need some policy regardless.

> I think I might prefer not accepting IL that carries byrefs across awaits. It feels a bit more deterministic, even if it may depend on whether the code was actually JIT-ed and perhaps on quality of flow analysis (and thus on JIT version or optimization mode).

It does seem hard to come up with good IL validity rules without having something like `HoistedLocal` (in my opinion we should call it `HoistableLocal` if we end up with it).

**Comment by VSadov** [comment]
this may no longer be relevant.
